.test-infra/jenkins/Flink.groovy

```diff
@@ -0,0 +1,114 @@
+  private static final String FLINK_DIR = '"$WORKSPACE/src/.test-infra/dataproc"'
+  private static final String FLINK_SCRIPT = 'flink_cluster.sh'
+  private def job
+  private String jobName
+  private String jobServerImageTag
+
+  Flink(job, String jobName) {
+    this.job = job
+    this.jobName = jobName
+    this.jobServerImageTag = ''
+  }
+
+  /**
+   * Creates Flink cluster and specifies cleanup steps.
+   *
+   * @param sdkHarnessImages - the list of published SDK Harness images tags
+   * @param workerCount - the initial number of worker nodes
+   * @param slotsPerTaskmanager - the number of slots per Flink task manager
+   */
+  void setUp(List<String> sdkHarnessImages, Integer workerCount, Integer slotsPerTaskmanager = 1) {
+    setupFlinkCluster(sdkHarnessImages, workerCount, slotsPerTaskmanager)
+    addTeardownFlinkStep()
+  }
+
+  /**
+   * Updates the number of worker nodes in a cluster.
+   *
+   * @param workerCount - the new number of worker nodes in the cluster
+   */
+  void scaleCluster(Integer workerCount) {
+    job.steps {
+      shell("echo Changing number of workers to ${workerCount}")
+      environmentVariables {
+        env("FLINK_NUM_WORKERS", workerCount)
+      }
+      shell("cd ${FLINK_DIR}; ./${FLINK_SCRIPT} scale")
+    }
+  }
+
+  void prepareJobServer(DockerPublisher jobServerPublisher = new DockerPublisher()) {
+    this.jobServerImageTag = jobServerPublisher.getFullImageName('flink-job-server')
+    jobServerPublisher.publish(job, ":runners:flink:${flinkVersion}:job-server-container:docker", 'flink-job-server')
+  }
+
+  private void setupFlinkCluster(List<String> sdkHarnessImages, Integer workerCount, Integer slotsPerTaskmanager) {
+    String gcsBucket = 'gs://beam-flink-cluster'
+    String clusterName = getClusterName()
+    String artifactsDir = "${gcsBucket}/${clusterName}"
+    String imagesToPull = sdkHarnessImages.join(' ')
+
+    job.steps {
+      environmentVariables {
+        env("GCLOUD_ZONE", "us-central1-a")
+        env("CLUSTER_NAME", clusterName)
+        env("GCS_BUCKET", gcsBucket)
+        env("FLINK_DOWNLOAD_URL", flinkDownloadUrl)
+        env("FLINK_NUM_WORKERS", workerCount)
+        env("FLINK_TASKMANAGER_SLOTS", slotsPerTaskmanager)
+        env("DETACHED_MODE", 'true')
+
+        if(imagesToPull) {
+          env("HARNESS_IMAGES_TO_PULL", imagesToPull)
+        }
+
+        if(jobServerImageTag) {
+          env("JOB_SERVER_IMAGE", jobServerImageTag)
+          env("ARTIFACTS_DIR", artifactsDir)
+        }
+      }
+
+      shell('echo Setting up flink cluster')
+      shell("cd ${FLINK_DIR}; ./${FLINK_SCRIPT} create")
+    }
+  }
+
+  private void addTeardownFlinkStep() {
+    job.publishers {
+      postBuildScripts {
+        steps {
+          shell("cd ${FLINK_DIR}; ./${FLINK_SCRIPT} delete")
+        }
+        onlyIfBuildSucceeds(false)
+        onlyIfBuildFails(false)
+      }
+    }
+  }
+
+  private GString getClusterName() {
```

Comment:
this should be below `setupFlinkCluster` because it's the only one method using this private method

Reply:
Totally agree. Thanks for spotting this.